File: aoc.cpp
#include <solution.hpp>

#include <fstream>
#include <iostream>
#include <filesystem>

namespace
{
    std::vector<std::string> readFileLines(std::filesystem::path path)
    {
        std::ifstream f(path);
        std::string buf;
        std::vector<std::string> result;
        while (std::getline(f, buf))
        {
            result.push_back(std::move(buf));
        }
        return result;
    }
}

int main(int argc, char *argv[])
{
    if (argc != 3)
    {
        std::cout << "Usage: aoc.cpp YEAR DAY\n";
        std::exit(1);
    }

    const std::string year = argv[1];
    const std::string day = argv[2];

    const auto part1_example_data(readFileLines(std::filesystem::path(year) / "data" / (day + ".example.txt")));
    const auto data(readFileLines(std::filesystem::path(year) / "data" / (day + ".txt")));

    std::vector<std::string> part2_example_data;
    const std::filesystem::path part2_example_path(std::filesystem::path(year) / "data" / (day + ".example2.txt"));
    if (std::filesystem::exists(part2_example_path))
    {
        part2_example_data = readFileLines(part2_example_path);
    }
    else
    {
        part2_example_data = std::move(part1_example_data);
    }

    const auto solutions = readFileLines(std::filesystem::path(year) / "data" / (day + ".solution.txt"));

    Solution s;
    const auto part1_example_result = s.part1(part1_example_data);
    if (part1_example_result != solutions[0])
    {
        std::cout << "Part 1 example failed, expected: " << solutions[0] << ", got: " << part1_example_result << std::endl;
        std::exit(1);
    }
    std::cout << "Part 1 example passed!" << std::endl;

    const auto part1_result = s.part1(data);
    std::cout << "Part 1 result: " << part1_result << std::endl;

    if (solutions.size() == 1)
    {
        std::cout << "Part 2 example solution not provided. Exiting." << std::endl;
        std::exit(0);
    }

    const auto part2_example_result = s.part2(part2_example_data);
    if (part2_example_result != solutions[1])
    {
        std::cout << "Part 2 example failed, expected: " << solutions[1] << ", got: " << part2_example_result << std::endl;
        std::exit(1);
    }
    std::cout << "Part 2 example passed!" << std::endl;

    const auto part2_result = s.part2(data);
    std::cout << "Part 2 result: " << part2_result << std::endl;
}

File: aoc.py
from pathlib import Path
import importlib
import sys
import time
import datetime

class Utils:
    def get_data(filename, year=datetime.date.today().year):
        return Utils.read_file_lines(Path('.') / str(year) / "data" / filename)
        
    def read_file_lines(file):
        with open(file) as f:
            return f.read().splitlines()

    def timed(func, *args, **kwargs):
        t = time.time()
        r = func(*args, **kwargs)
        t = time.time() - t
        return r, t

def process(year, day):
    module = importlib.import_module(year + '.' + day)

    input_lines = Utils.get_data(day + '.txt', year)
    example_input_lines = Utils.get_data(day + '.example.txt', year)

    part2_example_file = Path('.') / str(year) / "data" / f'{day}.example2.txt'
    part2_example_input_lines = Utils.read_file_lines(part2_example_file) if Path.exists(part2_example_file) else example_input_lines

    only_test = 'only_test' in dir(module)

    solution_file = Path('.') / str(year) / "data" / f'{day}.solution.txt'
    solutions = Utils.read_file_lines(solution_file) if Path.exists(solution_file) else module.example_answers

    if 'preprocess' in dir(module):
        module.preprocess(Utils)

    if 'skip_part1' not in dir(module):
        answer = str(module.part1(example_input_lines))
        assert answer == solutions[0], f"Part 1 failed! expected: {solutions[0]}, got: {answer}"
        if not only_test:
            print(module.part1(input_lines))

    if len(solutions) > 1:
        answer = str(module.part2(part2_example_input_lines))
        assert answer == solutions[1], f"Part 2 failed! expected: {solutions[1]}, got: {answer}"
        if not only_test:
            print(module.part2(input_lines))


if __name__ == "__main__":
    # Hack: you can pass either the full path of implementation file (easier to do in .vscode/launch.json), 
    #       or YEAR/DAY (if running manually)
    p = Path(sys.argv[1])
    process(p.parts[-2], p.parts[-1].partition('.')[0])

